subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -138,75 +136,82 @@ static int download_client_callback(const struct download_client_evt *event)
 			}
 
 			if (err_cause != FOTA_DOWNLOAD_ERROR_CAUSE_NO_ERROR) {
-				(void)download_client_disconnect(&dlc);
-				send_error_evt(err_cause);
-				int res = dfu_target_reset();
-
-				if (res != 0) {
-					LOG_ERR("Unable to reset DFU target, err: %d",
-						res);
-				}
-				first_fragment = true;
-				return err;
+				goto error_and_close;
 			}
 
 			err = dfu_target_offset_get(&offset);
 			if (err != 0) {
 				LOG_DBG("unable to get dfu target offset err: "
 					"%d", err);
-				send_error_evt(FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED);
-
-				return -1;
+				err_cause = FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED;
+				goto error_and_close;
 			}
 
 			if (offset != 0) {
-				/* Abort current download procedure, and
-				 * schedule new download from offset.
-				 */
-				(void)download_client_disconnect(&dlc);
-				k_work_schedule(&dlc_with_offset_work,
-						K_SECONDS(1));
-				LOG_INF("Refuse fragment, restart with offset");
-
-				return -1;
+				if (atomic_test_bit(&flags, FLAG_NEW_URI)) {
+					atomic_clear_bit(&flags, FLAG_RESUME);
+					/* Image is different, reset DFU target */
+					err = dfu_target_reset();
+
+					if (err != 0) {
+						LOG_ERR("Unable to reset DFU target, err: %d", err);
+						err_cause =
+							FOTA_DOWNLOAD_ERROR_CAUSE_DOWNLOAD_FAILED;
+						goto error_and_close;
+					}
+					err = dfu_target_init(img_type, 0, file_size,
+						      dfu_target_callback_handler);
+					if (err != 0) {
+						LOG_ERR("Failed to re-initialize target, err: %d",
+							err);
+						err_cause = FOTA_DOWNLOAD_ERROR_CAUSE_INTERNAL;
+						goto error_and_close;
+					}
+				} else {
+					/* Abort current download procedure, and
+					 * schedule new download from offset.
+					 */
+					atomic_set_bit(&flags, FLAG_RESUME);
+					(void)download_client_disconnect(&dlc);
+					k_work_schedule(&dlc_with_offset_work,
+							K_SECONDS(1));
+					LOG_INF("Refuse fragment, restart with offset");
+
+					return -1;
+				}
+			} else {
+				atomic_clear_bit(&flags, FLAG_RESUME);
 			}
 		}
 
 		err = dfu_target_write(event->fragment.buf,
 				       event->fragment.len);
 		if (err != 0) {
 			LOG_ERR("dfu_target_write error %d", err);
-			int res = dfu_target_done(false);
-
-			if (res != 0) {
-				LOG_ERR("Unable to free DFU target resources");
-			}
-			first_fragment = true;
-			(void) download_client_disconnect(&dlc);
-			send_error_evt(FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE);
```

Comment:
what is the reason for changing the error cause here?
`FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE` seems more descriptive and appropriate.
the CI is failing because the test/fota sample expect the error code to be `INVALID_UPDATE` when an invalid modem delta update is attempted.

Reply:
That is a valid question.
It looks like this is another undocumented feature. Modem Delta update responds `-EINVAL` when it rejects the delta image. All other cases, it is just the flash write have failed. I'll add a if-else for the corner case.